### DISCUSSION TITLE
MSBuild Android support

### DISCUSSION THREAD

**Discussion by sopyer** [microsoft/vcpkg#17812] [Ideas]
Add enable *-android targets integration with VS Android projects:

Here is simple patch which enabled compilation:

```diff
 scripts/buildsystems/msbuild/vcpkg.targets | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/scripts/buildsystems/msbuild/vcpkg.targets b/scripts/buildsystems/msbuild/vcpkg.targets
index 8807befd8..ddd3f6c0a 100644
--- a/scripts/buildsystems/msbuild/vcpkg.targets
+++ b/scripts/buildsystems/msbuild/vcpkg.targets
@@ -49,6 +49,30 @@
     <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">arm64-uwp</VcpkgTriplet>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'x86|Android'">
+    <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
+    <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x86-android</VcpkgTriplet>
+    <VcpkgAutoLink Condition="'$(VcpkgAutoLink)' == ''">false</VcpkgAutoLink>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'x64|Android|3.0'">
+    <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
+    <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">x64-android</VcpkgTriplet>
+    <VcpkgAutoLink Condition="'$(VcpkgAutoLink)' == ''">false</VcpkgAutoLink>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'ARM|Android|3.0'">
+    <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
+    <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">arm-android</VcpkgTriplet>
+    <VcpkgAutoLink Condition="'$(VcpkgAutoLink)' == ''">false</VcpkgAutoLink>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)' == 'ARM64|Android|3.0'">
+    <VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
+    <VcpkgTriplet Condition="'$(VcpkgTriplet)' == ''">arm64-android</VcpkgTriplet>
+    <VcpkgAutoLink Condition="'$(VcpkgAutoLink)' == ''">false</VcpkgAutoLink>
+  </PropertyGroup>
+
   <PropertyGroup Condition="'$(VcpkgEnabled)' == 'true'">
     <VcpkgConfiguration Condition="'$(VcpkgConfiguration)' == ''">$(Configuration)</VcpkgConfiguration>
     <VcpkgNormalizedConfiguration Condition="$(VcpkgConfiguration.StartsWith('Debug'))">Debug</VcpkgNormalizedConfiguration>

```

But I see following in Build console:
```
1>Target AppLocalFromInstalled:
1>  dumpbin : The term 'dumpbin' is not recognized as the name of a cmdlet, function, script file, or operable program.
1>  Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
1>  At D:\dev\vcpkg\scripts\buildsystems\msbuild\applocal.ps1:60 char:12
1>  +     $a = $(dumpbin /DEPENDENTS $targetBinary | ? { $_ -match "^    [^ ...
1>  +            ~~~~~~~
1>      + CategoryInfo          : ObjectNotFound: (dumpbin:String) [], CommandNotFoundException
1>      + FullyQualifiedErrorId : CommandNotFoundException
```

**Comment by MVoz** [comment]
`$(Platform)|$(ApplicationType)|$(ApplicationTypeRevision)`

![2020-05-27_041518](https://user-images.githubusercontent.com/10104740/82946449-f58e9e80-9fd0-11ea-82d8-e258fa527903.png)

______

do you have any experience in integrating this platform under cmake ?

**Comment by sopyer** [comment]
@Voskrese  no I don't, I have no experience in custom toolchains with MS toolsets. I am interested in using vcpkg built libraries in VS Android via vcxproj/msbuild at the moment, arm64/clang. This setup as I understand uses NDK compilation toolchain.

**Comment by MVoz** [comment]
strange, as far as I know, compilation under ANDROID in VCPKG does not affect MSBUILD
why then do these edits?

I thought that you have some experience using PLATFORMTOOLSET ANDROID MSBUILD

**Comment by sopyer** [comment]
This issue is not about Android compilation(worked out of the box for me). Its about vcpkg integration with VS/msbuild. I just reverse engineered those properties from msbuild output, trying to integrate.

**Comment by MVoz** [comment]
https://docs.microsoft.com/en-us/visualstudio/extensibility/visual-cpp-project-extensibility?view=vs-2019

https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_build_msbuild.cmake

https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_install_msbuild.cmake

**Comment by sopyer** [comment]
Please note changes are done to scripts/buildsystems/msbuild/vcpkg.targets.

This script is used for INTEGRATION.

Please see https://vcpkg.readthedocs.io/en/latest/about/faq/#i-cant-use-user-wide-integration-can-i-use-a-per-project-integration
As I understand it is installed when `vcpkg integrate install` is run. I tried my changes on project locally first in order to not break my system wide integration.

**Comment by MVoz** [comment]
> D:\dev\vcpkg\scripts\buildsystems\msbuild\ **applocal.ps1**
> $( **dumpbin.exe /DEPENDENTS** $targetBinary
> 
> this script with dumpbin looks for and copies dependencies

so did your edit help you avoid further errors and what was the port?

**Comment by MVoz** [comment]
you're writing code in MSVC for the ANDROID platform and trying to integrate VCPKG into your project, right?

open PR https://github.com/microsoft/vcpkg/pulls

**Comment by sopyer** [comment]
Yes this issue is about adding Visual Studio integration for Android projects. My patch is just demonstration that this is possible and should be relatively straightforward, but I am not an expert in MSBuild and vcpkg, and I had troubles with copying part - that is why I used static linkage in my triplets.
So I have opened issue to add that support by more knowledgeable engineer. And left my preliminary findings in order to simplify things.

**Comment by sopyer** [comment]
Another hint for person responsible for fix.
Thinking on discussion here, I checked global integration. And it does not work. Problem seems to be in MSBuild files.
It seems Microsoft.CppCommon.targets relies on Microsoft.Cpp.Common.props but it is not imported for Clang toolset. So my fix was to add `VCLibPackagePath` property to that file. After that I was able to import android libraries.
```xml
...
  <PropertyGroup>
    <VCLibPackagePath Condition="'$(VCLibPackagePath)' == ''">$(LOCALAPPDATA)\vcpkg\vcpkg.user</VCLibPackagePath>
  </PropertyGroup>
  <Import Condition="'$(VCLibPackagePath)' != '' and Exists('$(VCLibPackagePath).targets')" Project="$(VCLibPackagePath).targets" />
...
```
